src/pd/client.rs

```diff
@@ -62,40 +59,95 @@ fn send_msg(stream: &mut TcpStream, msg_id: u64, message: &Request) -> Result<(u
     Ok((id, resp.take_pd_resp()))
 }
 
-fn rpc_connect(endpoints: &str) -> Result<TcpStream> {
+fn rpc_connect(endpoint: &str) -> Result<TcpStream> {
+    let mut stream = box_try!(make_std_tcp_conn(endpoint));
+    try!(stream.set_write_timeout(Some(Duration::from_secs(SOCKET_WRITE_TIMEOUT))));
+
+    // Send a HTTP header to tell PD to hijack this connection for RPC.
+    let header_str = format!("GET {} HTTP/1.0\r\n\r\n", PD_RPC_PREFIX);
+    let header = header_str.as_bytes();
+    box_try!(stream.write_all(header));
+
+    Ok(stream)
+}
+
+// TODO: handle removed PD.
+fn validate_and_connect(endpoints: &[String], cluster_id: u64) -> Result<TcpStream> {
+    if endpoints.is_empty() {
+        return Err(box_err!("empty PD list"));
+    }
+
     // Randomize hosts.
-    let mut hosts: Vec<String> = endpoints.split(',').map(|s| s.into()).collect();
-    rand::thread_rng().shuffle(&mut hosts);
-
-    for host in &hosts {
-        let mut stream = match make_std_tcp_conn(host.as_str()) {
-            Ok(stream) => stream,
-            Err(_) => continue,
-        };
-        try!(stream.set_write_timeout(Some(Duration::from_secs(SOCKET_WRITE_TIMEOUT))));
-
-        // Send a HTTP header to tell PD to hijack this connection for RPC.
-        let header_str = format!("GET {} HTTP/1.0\r\n\r\n", PD_RPC_PREFIX);
-        let header = header_str.as_bytes();
-        match stream.write_all(header) {
-            Ok(_) => return Ok(stream),
-            Err(_) => continue,
+    let len = endpoints.len();
+    let mut indexes: Vec<usize> = (0..len).collect();
+    rand::thread_rng().shuffle(&mut indexes);
+
+    let mut pd_stream = None;
+    let mut endpoints_set = HashSet::with_capacity(len);
+    let mut members_resps = Vec::with_capacity(len);
+
+    for i in indexes {
+        let ep = &endpoints[i];
+        let mut stream = try!(rpc_connect(ep.as_str()));
+        let mut req = new_request(cluster_id, CommandType::GetPDMembers);
+        req.set_get_pd_members(GetPDMembersRequest::new());
+        let (id, mut resp) = try!(send_msg(&mut stream, VALIDATION_MSG_ID, &req));
```

Comment:
What if some nodes are down? For example, 1, 2, 3 are still online, but 4, 5 are down, then the validate_and_connect may not succeed.